lib/streamlit/runtime/caching/storage/cache_storage_protocol.py

```diff
@@ -0,0 +1,204 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from abc import abstractmethod
+from dataclasses import dataclass
+
+from typing_extensions import Literal, Protocol
+
+
+class CacheStorageError(Exception):
+    """Base exception raised by the cache storage"""
+
+
+class CacheStorageKeyNotFoundError(CacheStorageError):
+    """Raised when the key is not found in the cache storage"""
+
+
+class CacheStorageImproperlyConfigured(CacheStorageError):
+    """
+    Raised if the cache storage manager is not able to work with
+    provided CacheStorageContext.
+    """
+
+
+@dataclass(frozen=True)
+class CacheStorageContext:
+    """
+    Context passed to the cache storage during initialization
+    This is the normalized parameters that are passed to CacheStorageManager.create()
+    method.
+
+    Parameters
+    ----------
+    function_key: str
+        A hash computed based on function name and source code decorated
+        by st.cache_data
+
+    ttl_seconds : float or None
+        The time-to-live for the keys in storage, in seconds. If None, the entry
+        will never expire.
+
+    max_entries : int or None
+        The maximum number of entries to store in the cache storage.
+        If None, the cache storage will not limit the number of entries.
+
+    persist : Literal["disk"] or None
+        The persistence mode for the cache storage.
+        Legacy parameter, that used in Streamlit current cache storage implementation.
+        Could be ignored by cache storage implementation, if storage does not support
+        persistence or it persistent by default.
+    """
+
+    function_key: str
+    function_display_name: str
+    ttl_seconds: float | None = None
+    max_entries: int | None = None
+    persist: Literal["disk"] | None = None
+
+
+class CacheStorage(Protocol):
```

Comment:
I absolutely see your point, we definitely have too many layers of indirections for caching. 
In `cache_utils.py` we have `Cache`, `CachedFunc`, and `CachedFuncInfo` classes, as well as `DataCaches` and `DataCache` classes in `cache_data_api.py` to implement all machinery required for keeping the ability to decorate functions. 

That's the reason I tried to abstract the layer responsible for reading/writing results from that machinery of function decoration. This `CacheStorage` and `CacheStorageManager` was designed keeping the idea in mind that they should not know anything about what and how were caches, working purely with keys and bytes.

It wasn't 100% done, e.g. `CacheStorageContext` should know 
about `function_key` (to prefix files), and `function_display_name` to raise exceptions with that name, but at least our storage doesn't know anything about function entity, params hashing algorithms, etc. 

I agree that the motivation for why we have many CacheStorages per function instead of one should be better explained in the documentation/diagrams, but I want to keep the CacheStorage name as is since it should be not connected with the function.

Reply:
Seems reasonable -- I think my response in this case would be to push for a class diagram explaining how all these things work together so that it's more clear to the reader before they've read through all of the code + put some work into figuring it out